tests/h/security/test_permits.py

```diff
@@ -0,0 +1,73 @@
+from unittest.mock import sentinel
+
+import pytest
+
+from h.security import Identity, Permission
+from h.security.permits import identity_permits
+from h.traversal import AnnotationContext, Root
+
+
+class TestIdentityPermits:
+    def test_it(self, principals_for_identity, ACLAuthorizationPolicy):
+        result = identity_permits(
+            sentinel.identity, sentinel.context, sentinel.permission
+        )
+
+        principals_for_identity.assert_called_once_with(sentinel.identity)
+        ACLAuthorizationPolicy.assert_called_once_with()
+        ACLAuthorizationPolicy.return_value.permits.assert_called_once_with(
+            context=sentinel.context,
+            principals=principals_for_identity.return_value,
+            permission=sentinel.permission,
+        )
+        assert result == ACLAuthorizationPolicy.return_value.permits.return_value
+
+    @pytest.fixture
+    def principals_for_identity(self, patch):
+        return patch("h.security.permits.principals_for_identity")
+
+    @pytest.fixture
+    def ACLAuthorizationPolicy(self, patch):
+        return patch("h.security.permits.ACLAuthorizationPolicy")
+
+
+class TestIdentityPermitsIntegrated:
+    def test_it(self, user, group, annotation):
+        # We aren't going to go bonkers here, but a couple of tests to show
+        # this actually holds together. This isn't really to inform us of any
+        # particular failure, but just give us sensitivity if this doesn't work
+        # at all when you hook it together for real.
```

Comment:
I want these tests to help us bridge over the gap when we swap one system for another. Also the test above doesn't really prove anything.

Reply:
Hmmmm... I think I might want to avoid the two different classes here:

* Normally what we call a "unit test" at Hypothesis would not mock a third-party dependency like `ACLAuthorizationPolicy` (unless there was some good reason to). So we normally test "in integration with" third-party and stdlib dependencies but those are what we call "unit tests" not "integration tests"

* I can see that your `TestIdentityPermitsIntegrated` class is also not mocking our own `principals_for_identity()` function, so it's testing in integration with a function from another one of our own modules. In that sense it is actually different from what a Hypothesis "unit test" would normally do

* But to be honest doesn't `principals_for_identity()` belong here in the `permits.py` file anyway? We wouldn't normally mock another function from the same file (again, unless we had good reason to) so maybe just move it into this file

Anyway, I think I'd suggest:

1. Delete the `TestIdentityPermits` class
2. Rename `TestIdentityPermitsIntegrated` to `TestIdentityPermits`
3. If you think that the original `TestIdentityPermits.test_it()` method (the one with the mock `principals_for_identity()` and `ACLAuthorizationPolicy`) still has value then just keep it in the new `TestIdentifyPermits` class with the non-mocked tests

Reply:
Why mix the two things together?

Reply:
I don't think I really agree.

The first test mocks everything because it's convenient. It's a pure plumbing test and it's intended to be. See [What's our approach to mocking and not-mocking things in Python unit tests?](https://stackoverflow.com/c/hypothesis/questions/313)

> Guide: You should mock **_anything_** that makes your tests easier to read or write.

The second is specifically named to highlight that it is _not_ a conventional unit test and is separate from the first one. You say that this is different from a normal "unit test" but don't say if this is good bad or neutral. My assumption is you're mentioning it because you think different is bad. I think the changes suggested might make it so it's less obvious it's different maybe? This would cause a causal reader to not notice it's a more integrated style test. I don't see an advantage to that.

This test is written this way to make it 100% black box (in contrast to the first test). This is to allow the insides to be refactored completely in the coming PRs while giving some comfort it works.

`principals_for_identity` is separate from this behavior, and is used independently. This is one PR in long chain which is seeking to more and more limit the presence of principals in general. They will go soon.